crates/stages/src/stages/utils.rs

```diff
@@ -0,0 +1,238 @@
+
+    for (index, element) in collector.iter()?.enumerate() {
+        let (k, v) = element?;
+        let sharded_key = decode_key(k)?;
+        let new_list = BlockNumberList::decompress_owned(v)?;
+
+        if index > 0 && index % interval == 0 && total_entries > 100 {
+            info!(target: "sync::stages::index_history", progress = %format!("{:.2}%", (index as f64 / total_entries as f64) * 100.0), "Writing indices");
+        }
+
+        // AccountsHistory: `Address`.
+        // StorageHistory: `Address.StorageKey`.
+        let partial_key = get_partial(sharded_key);
+
+        if current_partial != partial_key {
+            // We have reached the end of this subset of keys so
+            // we need to flush its last indice shard.
+            load_indices(
+                &mut write_cursor,
+                current_partial,
+                &mut current_list,
+                &sharded_key_factory,
+                append_only,
+                LoadMode::Flush,
+            )?;
+
+            current_partial = partial_key;
+            current_list.clear();
+
+            // If it's not the first sync, there might an existing shard already, so we need to
+            // merge it with the one coming from the collector
+            if !append_only {
+                if let Some((_, last_database_shard)) =
+                    write_cursor.seek_exact(sharded_key_factory(current_partial, u64::MAX))?
+                {
+                    current_list.extend(last_database_shard.iter());
+                }
+            }
+        }
+
+        current_list.extend(new_list.iter());
+        load_indices(
+            &mut write_cursor,
+            current_partial,
+            &mut current_list,
+            &sharded_key_factory,
+            append_only,
+            LoadMode::KeepLast,
+        )?;
+    }
+
+    // There will be one remaining shard that needs to be flushed to DB.
+    load_indices(
+        &mut write_cursor,
+        current_partial,
+        &mut current_list,
+        &sharded_key_factory,
+        append_only,
+        LoadMode::Flush,
+    )?;
+
+    Ok(())
+}
+
+/// Shard and insert the indice list according to [`LoadMode`] and its length.
+pub(crate) fn load_indices<H, C, P>(
+    cursor: &mut C,
+    partial_key: P,
+    list: &mut Vec<BlockNumber>,
+    sharded_key_factory: &impl Fn(P, BlockNumber) -> <H as Table>::Key,
+    append_only: bool,
+    mode: LoadMode,
+) -> Result<(), StageError>
+where
+    C: DbCursorRO<H> + DbCursorRW<H>,
+    H: Table<Value = BlockNumberList>,
+    P: Copy,
+{
+    if list.len() > NUM_OF_INDICES_IN_SHARD || mode.is_flush() {
+        let chunks = list
+            .chunks(NUM_OF_INDICES_IN_SHARD)
+            .map(|chunks| chunks.to_vec())
+            .collect::<Vec<Vec<u64>>>();
+
+        let mut iter = chunks.into_iter().peekable();
+        while let Some(chunk) = iter.next() {
+            let mut highest = *chunk.last().expect("only exist when there's an indice");
```

Comment:
not sure what does it mean

```suggestion
            let mut highest = *chunk.last().expect("only exist when there's an index");
```